packages/core/content-releases/server/src/services/release.ts

```diff
@@ -161,6 +164,89 @@ const createReleaseService = ({ strapi }: { strapi: LoadedStrapi }) => ({
 
     return contentTypesMeta;
   },
+  async publish(releaseId: PublishRelease.Request['params']['id']) {
+    // We need to pass the type because entityService.findOne is not returning the correct type
+    const releaseWithPopulatedActionEntries = (await strapi.entityService.findOne(
+      RELEASE_MODEL_UID,
+      releaseId,
+      {
+        populate: {
+          actions: {
+            populate: {
+              entry: true,
+            },
+          },
+        },
+      }
+    )) as unknown as Release;
+
+    if (!releaseWithPopulatedActionEntries) {
+      throw new errors.NotFoundError(`No release found for id ${releaseId}`);
+    }
+
+    if (releaseWithPopulatedActionEntries.releasedAt) {
+      throw new errors.ValidationError('Release already published');
+    }
+
+    if (releaseWithPopulatedActionEntries.actions.length === 0) {
+      throw new errors.ValidationError('No entries to publish');
+    }
+
+    /**
+     * We separate publish and unpublish actions group by content type
+     */
+    const actions: {
+      [key: UID.ContentType]: {
+        publish: ReleaseAction['entry'][];
+        unpublish: ReleaseAction['entry'][];
+      };
+    } = {};
+    for (const action of releaseWithPopulatedActionEntries.actions) {
+      const contentTypeUid = action.contentType;
+
+      if (!actions[contentTypeUid]) {
+        actions[contentTypeUid] = {
+          publish: [],
+          unpublish: [],
+        };
+      }
+
+      if (action.type === 'publish') {
+        actions[contentTypeUid].publish.push(action.entry);
+      } else {
+        actions[contentTypeUid].unpublish.push(action.entry);
+      }
+    }
+
+    const entityManagerService = strapi.plugin('content-manager').service('entity-manager');
+
+    // Only publish the release if all action updates are applied successfully to their entry, otherwise leave everything as is
+    await strapi.db.transaction(async () => {
+      for (const contentTypeUid of Object.keys(actions)) {
+        const { publish, unpublish } = actions[contentTypeUid as UID.ContentType];
+
+        if (publish.length > 0) {
+          await entityManagerService.publishMany(publish, contentTypeUid);
+        }
+
+        if (unpublish.length > 0) {
+          await entityManagerService.unpublishMany(unpublish, contentTypeUid);
+        }
+      }
+    });
+
+    const release = await strapi.entityService.update(RELEASE_MODEL_UID, releaseId, {
+      data: {
+        /*
+         * The type returned from the entity service: Partial<Input<"plugin::content-releases.release">> looks like it's wrong
+         */
+        // @ts-expect-error see above
+        releasedAt: new Date(),
+      },
+    });
+
+    return release;
```

Comment:
> If everything is okay, it should return the release with the new publishedAt field and all the entries related should be published/unpublished

Did you want to return the entries that were involved in the release? I'm not sure if we should or not but just a heads up, currently the updated release is returned without anything populated.

Reply:
For now they are not needed, so I prefer keep it like this, we can add it later if we need it

Reply:
ok sounds good